File: jenfengchang/188.cpp
class Solution {
public:
    int maxProfit(int k, vector<int> &prices) {
        int len = prices.size();
        if (len<2) return 0;
        if (k>len/2){ // simple case
            int ans = 0;
            for (int i=1; i<len; ++i){
                ans += max(prices[i] - prices[i-1],0);
            }
            return ans;
        }
        int hold[k+1];
        int rele[k+1];
        for (int i=0;i<=k;++i){
            hold[i] = INT_MIN;
            rele[i] = 0;
        }
        int cur;
        for (int i=0; i<len; ++i){
            cur = prices[i];
            for(int j=k; j>0; --j){
                rele[j] = max(rele[j],hold[j] + cur);//
                hold[j] = max(hold[j],rele[j-1] - cur);
            }
            for(int j=k; j>0; --j){
                cout << rele[j] << " ";
            }   
            cout << endl;
            for(int j=k; j>0; --j){
                cout << hold[j] << " ";
            }   
            cout << endl;
        }
        return rele[k];
    }
};

File: ztuo/188.java
class Solution {
    public int maxProfit(int k, int[] prices) {
        if(k >= prices.length / 2) {
            return unlimited(prices);
        }
        int[][] sell = new int[k + 1][prices.length];
        for(int i = 1; i <= k; i++) {
            int buy =  - prices[0];
            for(int j = 1; j < prices.length; j++) {
                sell[i][j] = Math.max(buy + prices[j], sell[i][j-1]);
                buy =  Math.max(buy, sell[i-1][j-1] - prices[j]);
            }
        }
        return sell[k][prices.length - 1];
    }
    public int unlimited(int[] prices) {
        int profit = 0;
        for(int i = 1; i < prices.length; i++) {
            if(prices[i] > prices[i-1]) profit += prices[i] - prices[i-1];
        }
        return profit;
    }
}